### PR TITLE
added support for generation-based query and delete of blobs.

### PR DESCRIPTION
**Pull request by tsinha**
Small modifications to get_blob and delete_blob to handle generation ids when querying buckets that have versioning enabled.

### CODE REVIEW

**Comment by tsinha** [comment]
I signed the CLA as Vital Labs, Inc.

File: gcloud/storage/bucket.py [hunk 1]
```diff
@@ -222,14 +227,22 @@ def get_blob(self, blob_name, client=None):
...
         :rtype: :class:`gcloud.storage.blob.Blob` or None
         :returns: The blob object if it exists, otherwise None.
         """
         client = self._require_client(client)
         blob = Blob(bucket=self, name=blob_name)
         try:
+            path = blob.path
+            if generation is not None:
+                path = (blob.path + '?generation=' +
+                        quote(str(generation), safe=''))
```
Comment (dhermes) [inline gcloud/storage/bucket.py]: We should move `generation` onto `blob` so that `blob.path` picks it up if need be.

File: gcloud/storage/bucket.py [hunk 2]
```diff
@@ -214,6 +215,10 @@ def get_blob(self, blob_name, client=None):
           <Blob: my-bucket, /path/to/blob.txt>
           >>> print bucket.get_blob('/does-not-exist.txt')
           None
+          >>> print bucket.get_blob(
+          ... '/path/to/versioned_blob.txt',
+          ... generation=generation_id)
+          <Blob: my-bucket, /path/to/versioned_blob.txt>
```
Comment (dhermes) [inline gcloud/storage/bucket.py]: Thanks a lot for updating the docstring!

**Comment by dhermes** [comment]
@tsinha You can ignore the AppVeyor failure for `test_upload_from_file_w_bound_client_multipart`, we are tracking it on #1434.

Updating `Bucket.delete_blob` and `Bucket.get_blob` is insufficient here, `Blob.delete` and `Blob.exists` and `Blob.reload` may also need updating (or you may get that for free when updating `Blob.path`)

@jgeewax What should be done about matching his signature of the CLA with his GitHub handle?

File: gcloud/storage/bucket.py (hunk 1 above)
Comment (tseaver) [inline gcloud/storage/bucket.py]: That would require making it a ctor argment (which wouldn't be a bad idea, anyway).

Comment (dhermes) [inline gcloud/storage/bucket.py]: Not here it wouldn't, since the `blob` is constructed two lines above, we could use a `generation` setter on the blob.

Comment (tsinha) [inline gcloud/storage/bucket.py]: I did this originally.  But, after seeing how the code was coming together and thinking of the models a bit, I came to the conclusion that the bucket was a better place for this.  

Here is my reasoning; I'd love your feedback:

1) The blob object doesn't facilitate querying, rather it's a representation of what the storage system believes exists.  Allowing someone to define the generation of a blob, through a setter, seems to be counter to the purpose of the generation property on the blob.  That is, blob.generation is the canonical ID returned by the storage system for that blob.  Not something a user should be able to define.  

2) The blob.path is an imputed property and constructed from the only two key pieces of information needed to define a blob: bucket and name.  User-defined generation seems a bit superfluous, since it has no real meaning in the definition of a blob.

These two points led me to think of the generation as a query parameter better suited to the bucket API.

The problem with my approach now is that it disregards all other query parameters (ifGenerationNotMatch, ifMetagenerationNotMatch, etc.).  

I think the next step for this path of development would be a generator function, in the bucket, that uses user-specified query parameters and the blob.path to create the full API query URL.

Let me know your thoughts... if you would like the query param to be through a setter to be at the blob level, and corresponding blob.path construction to happen there, I can refactor.

**Comment by tsinha** [comment]
I'll look at the blob level methods and make sure everything works with the bucket level query, or refactor as you suggested if needed.  

I'll post an update shortly.

File: gcloud/storage/bucket.py (hunk 1 above)
Comment (dhermes) [inline gcloud/storage/bucket.py]: I totally get why you'd go that route. A little of where how we envision these working together may help. We consider `Bucket.get_blob` and `Bucket.delete_blob` to be tiny do-nothing wrappers and put all verbs (i.e. API methods) relating to a noun (here the noun is `Blob`) on the relevant object.

**Comment by tsinha** [comment]
@dhermes I looked over and fixed blob.exists and blob.delete.

Still unsure about adding generation param directly into blob.path

It doesn't seem as clean as having the api_request create the appropriate paths via blob.path and query_params at the bucket level.

Otherwise, I think this PR should be ready to go.

**Comment by dhermes** [comment]
@tsinha the generation is required to be in the path, since it uniquely defines the object if specified.

`Bucket.delete_blob` and `Bucket.get_blob` are meant to be "dumb" (i.e. they don't know how to do much) and the knowledge of how to make the request should live on `Blob`.

**Comment by tsinha** [comment]
@dhermes Thanks for the feedback.
Can you help me brainstorm how to reconcile the way in which [api_requests are made](https://github.com/GoogleCloudPlatform/gcloud-python/blob/master/gcloud/connection.py#L182) with embedding the generation parameter at the blob level?

This confounds [my current WIP](https://github.com/VitalLabs/gcloud-python/blob/8d92cf55264ad8d2295103c1b320ee8cbe56e7c1/gcloud/storage/bucket.py#L492), which will enable the renaming of blobs while preserving versions, since the 'copyTo' API end-point requires another query param ('sourceGeneration') to copy the right version.

If we move generation into blob.path, I'd end up with something like the following:

https://www.googleapis.com/storage/v1/b/sourceBucket/o/sourceObject?generation=XXX/copyTo/b/destinationBucket/o/destinationObject?generation=YYY?sourceGeneration=XXX

But I may not be thinking it through completely... any advice you have would be helpful.

**Comment by dhermes** [comment]
I think you've just thought about it more than I had :grinning: 

However, I'd still rather special case `Bucket.copy_blob` than special case every other place a generation might be used.

You could also leave `Blob.path` as-is and we could have another property that also contains identifying informating in query params, like `Blob.path_with_params` (or a name that is less hideous).

**Comment by tsinha** [comment]
@dhermes I've added a new blob.path_with_params function to alleviate some of the checking at the bucket level: [here](https://github.com/VitalLabs/gcloud-python/blob/master/gcloud/storage/blob.py#L137) and [here](https://github.com/VitalLabs/gcloud-python/blob/master/gcloud/storage/bucket.py#L243).

I've also got a working version of the copy_blob and rename_blob functions with versioning support [here](https://github.com/VitalLabs/gcloud-python/blob/master/gcloud/storage/bucket.py#L464) and [here](https://github.com/VitalLabs/gcloud-python/blob/master/gcloud/storage/bucket.py#L515).

I changed the API for copy_blob a little to fit into the existing rename_blob logic... that is, if versioned copy is enabled, copy_blob returns a tuple of two lists that contain the old and new blobs... rename_blobs takes the old_blobs list from this tuple and deletes them, and then returns the new_blobs list.

If versioning is not enabled, then copy_blobs's return parameters are unchanged from the existing API.

What do you think about this approach?  I can add testing around this if it looks reasonable.

**Comment by dhermes** [comment]
@tsinha Sorry for the review delay. I'm traveling and won't have a chance to look at this until next Monday. (I feel bad, sorry for the negative experience.)

**Comment by tsinha** [comment]
No worries... I look forward to your feedback next week. Safe travels.

File: gcloud/storage/blob.py [hunk 3]
```diff
@@ -133,6 +134,22 @@ def path(self):
...
+        :returns: the URL path to this blob and a dictionary with the
+                 generation that can be used in query_params for
+                 connection.api_request
+        """
+
+        params = {}
+        if self.generation is not None:
+            params = {'generation': self.generation}
+
+        return (self.path, params)
```
Comment (dhermes) [inline gcloud/storage/blob.py]: Parens are superflous

File: gcloud/storage/blob.py [hunk 4]
```diff
@@ -133,6 +134,22 @@ def path(self):
         return self.path_helper(self.bucket.path, self.name)
 
     @property
+    def path_with_params(self):
+        """Getter property for the URL path to this Blob, with version.
```
Comment (dhermes) [inline gcloud/storage/blob.py]: s/with version/with generation

File: gcloud/storage/blob.py (hunk 3 above)
Comment (dhermes) [inline gcloud/storage/blob.py]: Also, I expected this to just be a string, though I'll hold off on comments until I see how it's used below.

File: gcloud/storage/blob.py [hunk 5]
```diff
@@ -133,6 +134,22 @@ def path(self):
...
+
+        :rtype: tuple of ``path`` (a string) and ``params`` (a dictionary)
+        :returns: the URL path to this blob and a dictionary with the
+                 generation that can be used in query_params for
+                 connection.api_request
```
Comment (dhermes) [inline gcloud/storage/blob.py]: Indent 1 more space to align with `the` in `:returns: the`

File: gcloud/storage/blob.py [hunk 6]
```diff
@@ -241,6 +258,9 @@ def exists(self, client=None):
             # We only need the status code (200 or not) so we seek to
             # minimize the returned payload.
             query_params = {'fields': 'name'}
+            if self.generation is not None:
+                query_params['generation'] = self.generation
```
Comment (dhermes) [inline gcloud/storage/blob.py]: Why not use `self.path_with_params` here? In my head, that was the point.

File: gcloud/storage/blob.py [hunk 7]
```diff
@@ -266,7 +286,8 @@ def delete(self, client=None):
                  (propagated from
                  :meth:`gcloud.storage.bucket.Bucket.delete_blob`).
         """
-        return self.bucket.delete_blob(self.name, client=client)
+        return self.bucket.delete_blob(self.name, client=client,
+                                       generation=self.generation)
```
Comment (dhermes) [inline gcloud/storage/blob.py]: I had forgotten we give such little info to` Bucket.delete_blob` (i.e. it only has the blob's name, not the object itself).

File: gcloud/storage/bucket.py [hunk 8]
```diff
@@ -103,7 +103,7 @@ def client(self):
         """The client bound to this bucket."""
         return self._client
 
-    def blob(self, blob_name, chunk_size=None):
+    def blob(self, blob_name, chunk_size=None, generation=None):
```
Comment (dhermes) [inline gcloud/storage/bucket.py]: No need for this here. The `@setter` can handle it.

File: gcloud/storage/bucket.py [hunk 9]
```diff
@@ -222,14 +232,20 @@ def get_blob(self, blob_name, client=None):
...
+        :type generation: int
+        :param generation: Optional. The generation id to retrieve in a bucket
+                           that supports versioning.
+
         :rtype: :class:`gcloud.storage.blob.Blob` or None
         :returns: The blob object if it exists, otherwise None.
         """
         client = self._require_client(client)
-        blob = Blob(bucket=self, name=blob_name)
+        blob = Blob(bucket=self, name=blob_name, generation=generation)
```
Comment (dhermes) [inline gcloud/storage/bucket.py]: RE: Comment above, you can just use the setter:

```python
blob = Blob(bucket=self, name=blob_name)
blob.generation = generation
```

File: gcloud/storage/bucket.py (hunk 8 above)
Comment (dhermes) [inline gcloud/storage/bucket.py]: Reasoning: `generation` is a property of a blob, but there are too many properties to provide in the constructor

File: gcloud/storage/bucket.py [hunk 10]
```diff
@@ -392,10 +412,15 @@ def delete_blob(self, blob_name, client=None):
         """
         client = self._require_client(client)
         blob_path = Blob.path_helper(self.path, blob_name)
+        query_params = {}
+        if generation is not None:
+            query_params = {'generation': generation}
```
Comment (dhermes) [inline gcloud/storage/bucket.py]: Shouldn't this use the new method that includes the path params?

File: gcloud/storage/bucket.py [hunk 11]
```diff
@@ -423,7 +448,13 @@ def delete_blobs(self, blobs, on_error=None, client=None):
                 blob_name = blob
                 if not isinstance(blob_name, six.string_types):
                     blob_name = blob.name
-                self.delete_blob(blob_name, client=client)
+
+                generation = None
+                if hasattr(blob, 'generation'):
```
Comment (dhermes) [inline gcloud/storage/bucket.py]: `generation` is a property so `hasattr` is not needed. You can bunch this in with the `not isinstance(blob_name, six.string_types)` above (since that represents a `Blob` object)

File: gcloud/storage/bucket.py [hunk 12]
```diff
@@ -448,18 +479,38 @@ def copy_blob(self, blob, destination_bucket, new_name=None,
...
+                method='POST', path=api_path,
+                query_params={'sourceGeneration': old_blob.generation},
+                _target_object=new_blob)
+            new_blob._set_properties(copy_result)
+            new_blobs.append(new_blob)
+
+        if versions:
+            return (new_blobs, old_blobs)
+        else:
+            return new_blobs[0]
```
Comment (dhermes) [inline gcloud/storage/bucket.py]: Let's do this in another PR. Scope should not grow *during* review :grinning:

File: gcloud/storage/bucket.py [hunk 13]
```diff
@@ -484,11 +535,19 @@ def rename_blob(self, blob, new_name, client=None):
...
+        """
+        new_blobs, old_blobs = self.copy_blob(blob, self, new_name,
+                                              client=client, versions=True)
+        for old_blob in old_blobs:
+            old_blob.delete(client=client)
+
+        if len(new_blobs) == 1:
+            return new_blobs[0]
+        else:
+            return new_blobs
```
Comment (dhermes) [inline gcloud/storage/bucket.py]: Ditto here on tackling in another PR.

**Comment by tsinha** [comment]
@dhermes I have not forgotten about this.  Tied up with some other stuff and will get back to this ASAP.

**Comment by dhermes** [comment]
Thanks for the update

**Comment by tillahoffmann** [comment]
@dhermes, @tsinha, thanks for looking into this. Do you have any updates on when this PR might get merged?

**Comment by dhermes** [comment]
@tillahoffmann There have been lots of changes to `blob` since this PR was originally sent. Some code paths in the current release (`google-cloud-storage==1.2.0`) will "just work" for this feature.

Peeking at the [code][1], at least the `Blob.download_*` functions **will** incorporate the generation (but, as a hack, only if `media_link` isn't set on the `Blob` instance). As for [delete][2], it calls out to [`Blob.path_helper`][3] which **does not** incorporate the generation.

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/storage-1.2.0/storage/google/cloud/storage/blob.py#L382-L398
[2]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/storage-1.2.0/storage/google/cloud/storage/bucket.py#L417
[3]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/storage-1.2.0/storage/google/cloud/storage/blob.py#L193

**Comment by tillahoffmann** [comment]
@dhermes, thanks for the update. Do you have any recommendations on how to access md5 and crc32c hashes given a generation number from python?

**Comment by dhermes** [comment]
Sure thing. Here is a somewhat icky hack (but it'll work for now).

For a file with multiple versions:

```
$ gsutil ls -a gs://some-bucket
gs://some-bucket/file.txt#1498874876972049
gs://some-bucket/file.txt#1498875135704497
```

If you `reload()` it like normal, you'll see the latest version:

```python
>>> from google.cloud import storage
>>>
>>> client = storage.Client()
>>> bucket = client.bucket('some-bucket')
>>> blob_normal = bucket.blob('file.txt')
>>> blob_normal.reload()
>>> blob_normal._properties
{u'bucket': u'some-bucket',
 u'contentType': u'text/plain',
 u'crc32c': u'JcxYOQ==',
 u'etag': u'CLH7moCB59QCEAE=',
 u'generation': u'1498875135704497',
 u'id': u'some-bucket/file.txt/1498875135704497',
 u'kind': u'storage#object',
 u'md5Hash': u'F/IyVDfyzIWJW12F0xO6+Q==',
 u'mediaLink': u'https://www.googleapis.com/download/storage/v1/b/some-bucket/o/file.txt?generation=1498875135704497&alt=media',
 u'metageneration': u'1',
 u'name': u'file.txt',
 u'selfLink': u'https://www.googleapis.com/storage/v1/b/some-bucket/o/file.txt',
 u'size': u'1672',
 u'storageClass': u'MULTI_REGIONAL',
 u'timeCreated': u'2017-07-01T02:12:15.663Z',
 u'timeStorageClassUpdated': u'2017-07-01T02:12:15.663Z',
 u'updated': u'2017-07-01T02:12:15.663Z'}
```

However, if you manually patch `reload` on the object, you can get the data you want:

```python
>>> from google.cloud.storage.blob import _quote
>>>
>>> GENERATION = '1498874876972049'
>>>
>>> def reload(self, client=None):
...     client = self._require_client(client)
...     query_params = {'projection': 'noAcl', 'generation': GENERATION}
...     api_response = client._connection.api_request(
...         method='GET', path=self.path, query_params=query_params,
...         _target_object=self)
...     self._set_properties(api_response)
...
>>>
>>> blob_patched = bucket.blob('file.txt')
>>> blob_patched.reload = reload.__get__(blob_patched)
>>> blob_patched._properties
{u'bucket': u'some-bucket',
 u'contentType': u'text/plain',
 u'crc32c': u'MO9tZQ==',
 u'etag': u'CJGY64SA59QCEAE=',
 u'generation': u'1498874876972049',
 u'id': u'some-bucket/file.txt/1498874876972049',
 u'kind': u'storage#object',
 u'md5Hash': u'zPc9wCyNcKGiYV3ijiB2iw==',
 u'mediaLink': u'https://www.googleapis.com/download/storage/v1/b/some-bucket/o/file.txt?generation=1498874876972049&alt=media',
 u'metageneration': u'1',
 u'name': u'file.txt',
 u'selfLink': u'https://www.googleapis.com/storage/v1/b/some-bucket/o/file.txt',
 u'size': u'1635',
 u'storageClass': u'MULTI_REGIONAL',
 u'timeCreated': u'2017-07-01T02:07:56.965Z',
 u'timeDeleted': u'2017-07-01T02:12:15.704Z',
 u'timeStorageClassUpdated': u'2017-07-01T02:07:56.965Z',
 u'updated': u'2017-07-01T02:07:56.965Z'}
```

**Comment by lukesneeringer** [comment]
Hi @tillahoffmann,
Since the code has changed so much since your original PR, and it sounds like there is an existing path to do what you need now, I am going to close this. I am sorry that we sat on it for so long.

If there are any outstanding issues remaining, feel free to re-open (or open a new issue).
